### DISCUSSION TITLE
Dropping support for the plugin version of Timber

### DISCUSSION THREAD

**Discussion by Levdbas** [timber/timber#2804] [Announcements]
Hi from the development team of Timber!

We are in the [final stretch](https://github.com/timber/timber/issues) to release Timber 2.0.  🎉

Timber 2.0 is a complete overhaul to the latest coding standards, fully supports PHP 8.2 and overall is a great step forward to make interacting with the Timber API's a lot more consistent across all functionalities.

With that being said, we also decided to say farewell to the plugin version of Timber.

**What does this mean**
- Timber 2.0 will **not** be released as a plugin anymore, so if you want the shiny new stuff, you will have to upgrade to Timber 2.0 in the near future.
- Since Timber 2.0 will be our new main branch we will not actively maintain the v1 branch anymore and will not release updates for the plugin version.
- We will release one more plugin update soon that will add a notice in the plugin screen about to inform the Plugin users about this exact change.

**Why we are dropping plugin support**

There are many reasons why we have decided to not release Timber as a plugin anymore, but to highlight the main reasons:

- Under the current guidelines of the WordPress plugin theme, Timber wouldn't be allowed as a plugin.
- Plugins are meant to be upgradable. But since Timber 2 is a complete overhaul, clicking the "update" button without making the necessary theme changes would crash your website.
- Timber is meant for **developers** and thus we will always recommend modern development tools like Composer.

We fully understands that this change will come as a surprise to many of you. We understand that this will give you additional work, but we do think this is the best way moving forward. We hope we can count on your support on this. 

**How to switch to the Composer-based install**
If you are still using the Timber plugin we will encourage you to switch to the Composer-based install of Timber v1 as soon as possible.
We have an [extensive guide](https://timber.github.io/docs/v1/getting-started/switch-to-composer/) on how to switch from the plugin version to the composer based one.

If you have any questions, please let us know.

**Comment by jarednova** [comment]
Below is an extensive list with guides and the reasons why we are not going to release Timber 2 in plugin version anymore:
* Backstory: [Why we are dropping support for the plugin in the first place](https://github.com/timber/timber/pull/2005)
* GitHub issue: [Roadmap for Timber 2.0](https://github.com/timber/timber/issues/2741)
* COMING SOON: Guide: [How do I switch over from the plugin version to the Composer based version of Timber?](https://timber.github.io/docs/getting-started/switch-to-composer/)

**Comment by Levdbas** [comment]
And we are [live ](https://github.com/timber/timber/releases/tag/2.0.0)everyone!

**Reply by jarednova** [reply]
Woooooooohoooooooo!!!!!!

**Comment by Fordefound** [comment]
Can I ask, for those using the existing plugin of Timber v1 - will the plugin be deleted when the Timber plugin is no longer being supported (ie will the theming currently being used on sites disappear or crash if you have not moved to composer?)

**Reply by rmens** [reply]
If you installed a plug-in it won't magically delete itself from your server. So no sites will break.

It's up to the maintainers to remove Timer v1 from the WordPress plug-in directory. However it would be a bit too soon to do this imho. Even if they delete if from the WordPress plug-in directory, that won't delete it from your site.

**Reply by claudiugubernu** [reply]
I have a few sites using the Timber plugin version on WP and today i found the time to transition on my local env. one of them from the plugin to using composer. I can say it wasn't such a big task, but i guess that depends on your own implementation. So far all runs as before, but too early to really commit all this to the staging then live env. 

The rest of my sites that are deployed run fine with the plugin version of Timber so unless i delete the plugin they will work fine. I do have the latest version of the plugin somewhere safe just in case.

**Reply by Levdbas** [reply]
Hi @claudiugubernu , did you directly update to v2 or to did you move to the composer v1 version? I would recommend to move to the composer version and in the long run update to v2.

**Reply by claudiugubernu** [reply]
Hi @Levdbas , initially i have moved to composer v1, but after the comment spent a few hours and updated to v2 on a different branch. 

It was a bit of a search&replace but so far ( for the site i have chose to migrate and update to composer Timber v2) all works as before. 

Had some issues with CPT that mixes ACF with native thumbnail image but i managed to resolve that by doing the following:

changed this :
` {% include 'atoms/image.twig' with { 'image' : TimberImage(services_post) } %}`
to:
```
{% for services_post in services_posts %}
  {% set post = get_post(services_post.ID) %}
  {% services_post.title %}
  {% services_post.description%}
  {% include 'atoms/image.twig' with { 'image' : get_image(post.thumbnail) } %}

{% endfor %}
```

thinking of adding this into php where i set the `$context['services_posts]`;

Not brave enough to go live with all this so will test furthure on a staging env. but must reiterate this project is a brochoure car valeting site. Thank you

**Comment by Maffsie** [comment]
ah that's a little bothersome. I'm using Timber because I already had a blog theme written using Twig templating, and did not want to write any PHP in service of switching from my old blog engine to Wordpress. Since you state that Timber is for developers, is there an alternative for someone who just wants to specify how a website should be laid out without having to write anything other than HTML and CSS?

**Reply by jarednova** [reply]
@Maffsie you are good to keep doing exactly that. With the supported version of Timber (2.x) we will continue development, issue resolution and patches changes in Twig, WordPress and PHP versions — this is what we mean by "support."

Version 1.x and the deployment to WordPress.org still exist and we intend to keep them there as long as the good people at WordPress.org will allow it (indefinitely, as we understand) so you are good to keep using it for both existing and new sites.

**Comment by addedlovely** [comment]
The documentation page below ( linked from the plugins page) is a 404.
https://timber.github.io/docs/v2/getting-started/switch-to-composer/

Looks like it should be:
https://timber.github.io/docs/v1/getting-started/switch-to-composer/

**Reply by Levdbas** [reply]
Hi @addedlovely , thank you for the suggestion! @jarednova , I just committed the correct link in [wp-plugin](https://github.com/timber/timber/tree/wp-plugin) branch. Could you push the readme.txt file to SVN?

**Reply by jarednova** [reply]
This is now complete. Thanks so much for the fix @addedlovely + @Levdbas !

**Comment by jarednova** [comment]
Sure thing! will take care of that later today

On Wed, Jun 5, 2024 at 2:20 AM Erik van der Bas ***@***.***>
wrote:

> Hi @addedlovely <https://github.com/addedlovely> , thank you for the
> suggestion! @jarednova <https://github.com/jarednova> , I just committed
> the correct link in wp-plugin
> <https://github.com/timber/timber/tree/wp-plugin> branch. Could you push
> the readme.txt file to SVN?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/timber/timber/discussions/2804#discussioncomment-9673683>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAJ45JUMNSGXUCIWTGHA7JLZF2U4PAVCNFSM6AAAAAA5CZKFC6VHI2DSMVQWIX3LMV43SRDJONRXK43TNFXW4Q3PNVWWK3TUHM4TMNZTGY4DG>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>


-- 

Jared Novack

Co-Founder, CFO & COO

***@***.***


<https://www.upstatement.com/?ref=email-signature>

**Reply by jarednova** [reply]
... this is now complete ✅

**Comment by jordane2086** [comment]
How do I update Timber? I just inherited a website to manage and it is not PHP 8.1+ compatible and breaking due to this plugin not being the new Composer version. Any guidance ?

**Reply by Levdbas** [reply]
See https://timber.github.io/docs/v1/getting-started/switch-to-composer/ and https://timber.github.io/docs/v2/upgrade-guides/2.0/ to update to the latest version.

**Reply by jordane2086** [reply]
Thanks. I am not a developer and running those codes it outside what I know to get that done. How long would it take someone to get those updated?

**Reply by Levdbas** [reply]
Really depends on the size of the website, how old the codebase is, how much the site relies on Timber.
My projects usually take around 5 hours up to two day to update.

**Comment by aladarmezga615** [comment]
**Is there a better approach to upgrading to Timber 2?**

We got the task of updating Timber 1 to Timber 2 in a legacy theme (10 years old). There are many Twig templates and PHP files that should be changed in the theme. We do not have the capacity for such a huge change, so we started searching for other solutions.

As we have many descendant classes of `Timber\Post` and we instantiate them through their constructors the `Timber::get_post()`-like instantiation would increase the amount of the to be changed files. Furthermore, our solution highly depends on the presence of the ACF fields on the post instance, as Timber 1 provided them earlier. To avoid the enormous changes in the existing code, we decided to extend the `Timber\Post` class keeping the ability of the usage of constructors and providing the ACF fields.

The constructor fetches and attaches the ACF fields to the corresponding instance. We did the same thing in the case of Terms.

We chose this solution because we have to focus on other more important problems and now we do not have the capacity to rewrite the existing code. We need Timber 2 to be in place implement new features with it and rewrite the old code when it needs to be changed.

And the question:

**What concerns can be seen in this solution?**

**Could other approaches and solutions be mentioned?**

Our main goal here is to have Timber 2 running and keeping the number of changes as low as possible.

Below is the custom Post class that became the new ancestor of our other classes. This way we needed to change extends `\Timber\Post` to extends` \Custom\Theme\Timber\Post` and everything started working.

```
namespace Custom\Theme\Timber;

use Timber\Timber;

class Post extends \Timber\Post
{
    public function __construct($pid = null)
    {
        $data= Timber::get_post($pid);
        $customs = $this->get_post_custom($data->id);
        $this->import(array_merge((array) $data, (array) $customs, ['custom' => $customs]));
        $this->wp_object = $data->wp_object;
    }

    public static function build(\WP_Post $wp_post): static
    {
        $post = parent::build($wp_post);
        $customs = $post->get_post_custom($post->id);
        $post->import(array_merge((array) $customs, ['custom' => $customs]));
        return $post;
    }

    protected function get_post_custom($pid)
    {
        $post_meta = get_post_meta($pid);
        foreach ($post_meta as $key => $value) {
            if (is_array($value) && count($value) == 1 && isset($value[0])) {
                $value = $value[0];
            }
            $post_meta[$key] = maybe_unserialize($value);
        }
        return $post_meta;
    }
}
```

**Reply by Levdbas** [reply]
You could use the `timber/post/import_data` instead to import the data on the build method. 
Use the -`>realize()` or -`>to_array()`  functionality to directly instantiate the post with all the data. This is probably the solution to directly access all the post data instead of lazily. Read more about this here: https://timber.github.io/docs/v2/guides/posts/#post-collections

**Reply by gchtr** [reply]
> Furthermore, our solution highly depends on the presence of the ACF fields on the post instance, as Timber 1 provided them earlier. To avoid the enormous changes in the existing code, we decided to extend the Timber\Post class keeping the ability of the usage of constructors and providing the ACF fields.

This seems like a reasonable approach for your use case.

Upgrading to Timber 2 doesn’t necessarily mean you have to do everything in a way that is proposed in the documentation.

In Timber 2, we tried to make it possible (and maybe easier) to extend Timber for use cases like yours. So if you load all custom fields into a property `custom`, then that’s totally fine and should also work in the future.

I’d say eventually, it would be better if you used `Timber::get_post()` instead of directly instantiating posts and worked with a [Class Map](https://timber.github.io/docs/v2/guides/class-maps/) to select the right class for your post to be built with. Because changing the visibility of `__construct()` to `public` in your child class is not a best coding practice.

But, you chose a compromise here, which should work for now.

> **Could other approaches and solutions be mentioned?**

Other than what @Levdbas mentioned, I wonder whether you could ditch your custom `get_post_custom()` method and use `meta()` instead. If you don’t pass a field name to `meta()`, it should all fields.

And maybe you would have to enable transforms as well, see https://timber.github.io/docs/v2/hooks/filters/#timber/meta/transform_value.
